metadata/private/rules/package_metadata.bzl

```diff
@@ -0,0 +1,86 @@
+"""Declares rule `package_metadata`."""
+
+load("//private/providers:package_attribute_info.bzl", "PackageAttributeInfo")
+load("//private/providers:package_metadata_info.bzl", "PackageMetadataInfo")
+
+def _package_metadata_impl(ctx):
+    attributes = [a[PackageAttributeInfo] for a in ctx.attr.attributes]
+
+    metadata = ctx.actions.declare_file("{}.package-metadata.json".format(ctx.attr.name))
+
+    ctx.actions.write(
+        output = metadata,
+        content = json.encode({
+            "attributes": {a.kind: a.metadata.path for a in attributes},
+            "label": str(ctx.label),
+            "purl": ctx.attr.purl,
+        }),
+    )
+
+    return [
+        DefaultInfo(
+            files = depset(
+                direct = [
+                    metadata,
+                ],
+            ),
+        ),
+        PackageMetadataInfo(
+            metadata = metadata,
+            files = depset(
+                direct = [
+                    metadata,
+                ],
+                transitive = [a.files for a in attributes],
```

Comment:
Should this also propagate the `a.metadata` file, or is it expected that `a.files` must contain `a.metadata`? If the latter, maybe document that assumption in the `PackageAttributeInfo.files` Provider docs.

Reply:
It's expected that `files` always contains (self) `metadata` file. I was planning to create a Starlark constructor function for `PackageAttributeInfo` that takes care of making sure `metadata` is always part of `files` (but haven't done in the PR to keep it as small as possible). I'll add something to the doc string